### ISSUE TITLE
[$500] Android - IOU-From BNP page, tap next and navigate back, phone keypad opens

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#30511]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.92-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:** @

## Action Performed:
1. Launch app
2. Tap fab
3. Tap request money
4. Enter an amount
5. Tap next
6. Tap app's back button

## Expected Result:
When user taps app's back button, phone keypad must not open.

## Actual Result:
When user taps app's back button, phone keypad opens.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/115492554/270dd1f6-7c8a-43d6-972b-02c07ed6833c

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c70a484f8520ecd4</li>
        <li>Upwork Job ID: 1717893710590554112</li>
        <li>Last Price Increase: 2024-01-02</li>
    </ul>
</details>

**Comment by manlaig** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue.
Keyboard is focused on Android when navigating back to screen.

## What is the root cause of that problem?
When `disableKeyboard` prop is enabled on `TextInput`, an AppState listener dismisses the keyboard on `background` or `inactive` events. It doesn't dismiss the keyboard on `active` event.

https://github.com/Expensify/App/blob/22678a5d0198e5281ede2c7b7e8b84d93abb6dd8/src/components/TextInput/index.native.js#L9-L19

## What changes do you think we should make in order to solve the problem?
Remove the conditional check inside the event callback that prevents dismissing the keyboard on `active` event. So, the listener would dismiss the keyboard on all AppState event changes and also initially dismiss the keyboard.
The code would look like:

```
    if (!props.disableKeyboard) {
        return;
    }
    Keyboard.dismiss();

    const appStateSubscription = AppState.addEventListener('change', (nextAppState) => {
        Keyboard.dismiss();
    });
```

## What alternative solutions did you explore? (Optional)
Since we don't need the keyboard when `disableKeyboard` is set, removing the conditional check simplifies the code and initially calling dismiss prevents bugs like this. I think this is a general solution that will address similar keyboard bugs in the future.

**Comment by wlegolas** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Android - IOU-From BNP page, tap next and navigate back, phone keypad opens

### What is the root cause of that problem?

When the user clicks on the "**Next**" button to select the participants, when the participants' list is shown by default the keyboard is displayed.

As the user clicks on the back button and this keyboard is with the state opened, when the previous view is shown the keyboard still opens. 

### What changes do you think we should make in order to solve the problem?

We need to change the Native input to add the logic to subscribe to the `blur` event from the `Navigation` and when the `blur` event occurs we need to call `Keyboard.dismiss()` to close the keyboard and change its state to close.

For example:

```
const TextInput = forwardRef((props, ref) => {
    const navigation = useNavigation();

    useEffect(() => {
        const unsubscribe = navigation.addListener('blur', () => {
            Keyboard.dismiss();
        });

        return unsubscribe;
    }, [navigation]);

    ...
});
```

### What alternative solutions did you explore? (Optional)

N/A

**POC**

https://github.com/Expensify/App/assets/698363/d36d1883-1844-47de-89cd-c2e5126cc77e

**Comment by s77rt** [comment]
Is there any specific requirements to reproduce this bug? Works fine from my end


https://github.com/Expensify/App/assets/16493223/de0a569c-25c7-422d-ab8f-66631e41cdd6

**Comment by conorpendergrast** [comment]
I can't reproduce via Browserstack either, I get the regular in-app keyboard: 

<img width="317" alt="image" src="https://github.com/Expensify/App/assets/5802477/dc278a44-dabe-4157-8d78-2a4724c2afb7">

**Comment by conorpendergrast** [comment]
@izarutskaya could you confirm if there's any extra information needed for reproduction? Is this specific to one particular Android device?

**Comment by izarutskaya** [comment]
I can't reproduce this anymore. Build - 1.3.95-2

https://github.com/Expensify/App/assets/115492554/72a01c82-7b18-4f51-b79b-8e80bf6f532d

**Comment by s77rt** [comment]
I think we can close this @conorpendergrast

**Comment by conorpendergrast** [comment]
I agree, based on none of us being able to reproduce any more. Thanks for raising this @izarutskaya !

**Comment by lanitochka17** [comment]
Issue reproducible on latest build 1.4.1-5

Issue found when executing PR https://github.com/Expensify/App/pull/30873

https://github.com/Expensify/App/assets/78819774/c8c7c8ac-6be6-41f6-8c86-bcc1100fe159

![viber_image_2023-11-20_12-19-59-859 (1)](https://github.com/Expensify/App/assets/78819774/c2279b15-b215-495c-ac65-badcd58cd8dc)

**Comment by s77rt** [comment]
I can reproduce this but need to click the back button quick


https://github.com/Expensify/App/assets/16493223/47c6c2b0-ea97-4f15-ab52-ee3e6a2df3f8

**Comment by s77rt** [comment]
@manlaig Thanks for the proposal. I don't think your RCA is correct. That hook is used to solve another problem and have no effect in this issue.

**Comment by s77rt** [comment]
@wlegolas Thanks for the proposal. I think your RCA is incomplete. If you follow the reproduction steps slowly you won't encounter the issue.

> As the user clicks on the back button and this keyboard is with the state opened, when the previous view is shown the keyboard still opens.

This is not exactly correct


https://github.com/Expensify/App/assets/16493223/651d9ec9-bcc6-4a0d-b59d-36a35b874105

**Comment by wlegolas** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/30511#issuecomment-1783367895)

**Comment by wlegolas** [comment]
> @wlegolas Thanks for the proposal. I think your RCA is incomplete. If you follow the reproduction steps slowly you won't encounter the issue.
> 
> > As the user clicks on the back button and this keyboard is with the state opened, when the previous view is shown the keyboard still opens.
> 
> This is not exactly correct
> 
>  Screen.Recording.2023-11-20.at.8.23.30.PM.mov

Hi @s77rt thank you for reviewing my proposal and for your feedback about this proposal.

**Comment by s77rt** [comment]
@wlegolas Thanks for the update. I think the solution is a workaround. It feels like this kind of a bug should be fixed at a lower level e.g. ReactNative or Android

**Comment by wlegolas** [comment]
Hi @s77rt you're welcome.

I understood your points to fix this issue at a lower level.

If there is anything else I can help with, please let me know.

Thank you for your support in this issue @s77rt

**Comment by s77rt** [comment]
@conorpendergrast Can we bring this back External?

**Comment by conorpendergrast** [comment]
Yep, done

**Comment by situchan** [comment]
Maybe hold this for [money request flow refactor](https://github.com/Expensify/App/pull/28618)

**Comment by s77rt** [comment]
@bernhardoj Thanks for the proposal. Your RCA makes sense but the solution is only a workaround.

**Comment by conorpendergrast** [comment]
Looks like we're still looking for non-workaround proposals for this!

**Comment by s77rt** [comment]
Not overdue. Looking for more proposals. This probably needs to be fixed on Android

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Still looking for proposals

**Comment by s77rt** [comment]
@bernhardoj I think this should be fixed on Android. Thought I'm not sure how we can contribute to it. This issue does not seem to be related to the App itself so any solution would have to be a workaround.

**Comment by s77rt** [comment]
I think this is a problem at Android level. When switching between inputs the keyboard is kept open to avoid the extra animations. But in this case we should actually dismiss the keyboard.

**Comment by s77rt** [comment]
Not overdue. Still not sure how to proceed though, it's either we fix this at Android or close the issue since we can't do much about it (/App perspective)

**Comment by s77rt** [comment]
Asking in Slack if this can be closed https://expensify.slack.com/archives/C01GTK53T8Q/p1703083456350169 (low priority)

**Comment by s77rt** [comment]
Not overdue. Waiting on Slack input

**Comment by s77rt** [comment]
Same ^

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by s77rt** [comment]
Not overdue. Not expecting much in Slack thread this week. Will bump the thread next week

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Bumped the slack thread

**Comment by conorpendergrast** [comment]
I've replied [in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1704810011103159?thread_ts=1703083456.350169&cid=C01GTK53T8Q), sorry that took so long! Based on that, I'll close out. If any disagrees, add a comment!
